pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net"
+	"testing"
+)
+
+func initTestIPAllocator() *IPAllocator {
+	pendingIPRanges := make(map[string]bool)
+	return &IPAllocator{
+		pendingIPRanges: pendingIPRanges,
+	}
+}
+
+func TestValidateCIDR(t *testing.T) {
+	cases := []struct {
+		name     string
+		cidr     string
+		expected bool
+	}{
+		{
+			name:     "Valid /29 CIDR",
+			cidr:     "192.168.92.192/29",
+			expected: true,
+		},
+		{
+			name:     "Invalid CIDR",
+			cidr:     "192.168.92.192",
+			expected: false,
+		},
+		{
+			name:     "Valid /29 CIDR",
+			cidr:     "192.168.92.248/29",
+			expected: true,
+		}, {
+			name:     "Invalid CIDR with IP range overflow",
+			cidr:     "192.168.92.249/29",
+			expected: false,
+		},
+	}
+
+	ipAllocator := initTestIPAllocator()
+	for _, test := range cases {
+		validCIDR := ipAllocator.ValidateCIDR(test.cidr)
+		if validCIDR != test.expected {
+			t.Errorf("error while validating cidr %s, expected %t, actual %t", test.cidr, test.expected, validCIDR)
+		}
+	}
+}
+
+func TestGetUnReservedIPRange(t *testing.T) {
+	// Using IPs which are not the beginning IPs of /29 CIDRs to evaluate the edge case
+	ips := [8]string{"192.168.92.3/29", "192.168.92.10/29", "192.168.92.20/29", "192.168.92.28/29"}
+	cidr := "192.168.92.0/27"
+	cases := []struct {
+		name                          string
+		expected                      string
+		pendingIPRanges               map[string]bool
+		cloudProviderReservedIPRanges map[string]bool
+		errorExpected                 bool
+	}{
+		{
```

Comment:
some more test cases:
- 0 pending, 0/1 used
- pending and used ips are out of the cidr range

Reply:
also a test case that involves carryover

Reply:
Due to bitwise AND there would never be a carryover. Do you have an example?

Reply:
cidr = "192.168.0.0/16", expected ip range = "192.168.1.0/29"